pkg/util/k8scmcache.go

```diff
@@ -0,0 +1,208 @@
+                        os.Exit(1)
+                }
+        }
+        client, err := k8s.NewForConfig(cfg)
+        if err != nil {
+                glog.Errorf("Failed to create client with error: %v\n", err)
+                os.Exit(1)
+        }
+        return client
+}
+
+func (k8scm *K8sCMCache) createMetadataCM(resourceID string) (*v1.ConfigMap, error) {
+        cm := &v1.ConfigMap{}
+        cm, err := k8scm.getMetadataCM(resourceID)
+        if err != nil {
+                glog.Infof("k8s-cm-cache: couldn't get configmap %s with error: %v", resourceID, err)
+                glog.Infof("k8s-cm-cache: creating configmap...")
+                cm = &v1.ConfigMap {
+                        ObjectMeta: metav1.ObjectMeta {
+                                Name: resourceID,
+                                Namespace: k8scm.Namespace,
+                                Labels: map[string]string{
+                                        csiMetadataLabelAttr: cmLabel,
+                                },
+                        },
+                        Data: map[string]string{},
+                }
+                cm, err = k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).Create(cm)
+                if err != nil {
+                        glog.Errorf("k8s-cm-cache: couldn't create configmap %s with error: %v\n", resourceID, err)
+                        return nil, errors.Wrap(err, "k8s-cm-cache: create configmap error")
+                }
+        }
+        glog.Infof("k8s-cm-cache: configmap %s successfully created\n", resourceID)
+        return cm, nil
+}
+
+func (k8scm *K8sCMCache) getMetadataCM(resourceID string) (*v1.ConfigMap, error) {
+	k8sCMCacheMtx.Lock()
+	defer k8sCMCacheMtx.Unlock()
+
+        cm, err := k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).Get(resourceID, metav1.GetOptions{})
+        if err != nil {
+                return nil, err
+        }
+        return cm, nil
+}
+
+
+func (k8scm *K8sCMCache) ForAll(pattern string, destObj interface{}, f ForAllFunc) error {
+	listOpts := metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", csiMetadataLabelAttr, cmLabel)}
+	cms, err := k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).List(listOpts)
+        if err != nil {
+                glog.Infof("k8s-cm-cache: failed to list metadata configmaps")
+                return err
+        }
+
+	k8sCMCacheMtx.Lock()
+	defer k8sCMCacheMtx.Unlock()
+
+	for _, cm := range cms.Items {
+		data := cm.Data[cmDataKey]
+		match, err := regexp.MatchString(pattern, cm.ObjectMeta.Name)
+		if err != nil {
+			continue
+		}
+		if !match {
+			continue
+		}
+		if err := json.Unmarshal([]byte(data), destObj); err != nil {
+			glog.Infof("k8s-cm-cache: unmarshal error")
+			continue
+		}
+		if err = f(cm.ObjectMeta.Name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k8scm *K8sCMCache) Create(identifier string, data interface{}) error {
```

Comment:
What are the use-cases for overwriting an existing cache entry? IMHO this is out-of-scope for `Create` and I'd say if an entry with this identifier already exists just exit with success, otherwise create a new CM object (and return success in case of an "ALREADY EXISTS" error). That way you also get rid of locking explicitly.

Reply:
Still a pending issue. Is there a reason why the CM is being updated here? If so please post a link to the code, maybe I'm missing something. This method should only:
1. Try to `Get` the CM object (this is to handle idempotency)
1.1 If it already exists, `return nil` as to say everything is ok
2. Otherwise, `Create` a new CM object
2.2 If it already exists, also return success